src/input/reader_internal.cc

```diff
@@ -0,0 +1,820 @@
+        if (string_type != NULL ) return make_sub_storage(p, string_type );
+
+        // default -> error
+        THROW( Type::ExcUnknownDescendant() << Type::EI_TypeName(typeid(type).name()) );
+    }
+
+    return new StorageNull();
+}
+
+StorageBase * ReaderInternalBase::make_sub_storage(PathBase &p, const Type::Record *record)
+{
+	// control test, check correct tag (or TYPE key) if Record is derived from Abstract
+	string record_name_from_tag = p.get_record_tag();
+	if (record_name_from_tag == "include") {
+		return make_include_storage(p, record);
+	} else if (record_name_from_tag == "include_csv") {
+		THROW( ReaderToStorage::ExcForbiddenTag() << ReaderToStorage::EI_Tag("include_csv")
+			<< ReaderToStorage::EI_Specification("can be used only with arrays.") );
+	} else {
+		if ( record_name_from_tag != "" ) {
+			ASSERT(record_name_from_tag == record->type_name())(record_name_from_tag)(record->type_name()).error("Inconsistent tag of record.");
+		}
+		std::set<string> keys_to_process;
+		bool effectively_null = p.is_effectively_null();
+		if ( p.get_record_key_set(keys_to_process) || effectively_null ) {
+	        std::set<string>::iterator set_it;
+
+	        /*Type::Record::KeyIter key_it;
+	        if ( record->has_key_iterator("TYPE", key_it) && record->auto_conversion_key_iter() != record->end() ) {
+	            PathBase *type_path = p->clone();
+	            if ( type_path.down( "TYPE" ) ) {
+	                try {
+	                	ASSERT( type_path.get_string_value() == record->type_name() )(type_path.get_string_value())(record->type_name())
+	                		.error("Invalid value of TYPE key of record");
+	                    make_storage(type_path, key_it->type_.get() )->get_int();
+	                } catch(Type::Selection::ExcSelectionKeyNotFound &e) {
+	                	return record_automatic_conversion(p, record);
+	                }
+	            }
+	            else {  // automatic conversion
+	            	return record_automatic_conversion(p, record);
+	            }
+	        }*/
+
+	        StorageArray *storage_array = new StorageArray(record->size());
+	        // check individual keys
+	        for( Type::Record::KeyIter it= record->begin(); it != record->end(); ++it) {
+	        	// remove processed key from keys_to_process
+	        	set_it = keys_to_process.find(it->key_);
+	        	if (set_it != keys_to_process.end()) {
+	        		keys_to_process.erase(set_it);
+	        	}
+
+	            if ( !effectively_null && p.down(it->key_) ) {
+	                // key on input => check & use it
+	                // check for obsolete key
+
+	                auto obsolete_it = it->attributes.find( Type::Attribute::obsolete() );
+	                if ( obsolete_it != it->attributes.end()) {
+	                    WarningOut() << "Usage of the obsolete key: '" << it->key_ << "'\n" << obsolete_it -> second;
+	                }
+
+	                this->push_storage_index_to_vector(it->key_index);
```

Comment:
Move this functionality to Path classes so the StorageAddress will be available for all ReaderInternal implementations, but used just in CSV. Path.get_index_address() ... comment its functionality.